File: Src/Icm.Core/Tree/ITreeElementExtensions.vb
Imports System.Runtime.CompilerServices

Namespace Icm.Tree

    ''' <summary>
    ''' Tree traversals.
    ''' </summary>
    ''' <remarks>
    ''' 
    ''' </remarks>
    Public Module ITreeElementExtensions


        <Extension>
        Public Iterator Function DepthPreorderTraverse(Of T)(tn As ITreeElement(Of T)) As IEnumerable(Of T)
            Dim childStack As New Stack(Of IEnumerator(Of ITreeElement(Of T)))

            childStack.Push(tn.GetChildElements.GetEnumerator)
            Yield tn.Value
            Do Until childStack.Count = 0
                Dim childEnum = childStack.Peek

                If childEnum.MoveNext() Then
                    Dim child = childEnum.Current
                    childStack.Push(child.GetChildElements.GetEnumerator)
                    Yield child.Value
                Else
                    childStack.Pop()
                End If
            Loop
        End Function

        <Extension>
        Public Iterator Function DepthPostorderTraverse(Of T)(tn As ITreeElement(Of T)) As IEnumerable(Of T)
            Dim childStack As New Stack(Of IEnumerator(Of ITreeElement(Of T)))
            Dim rootEnum = {tn}.ToList.GetEnumerator
            childStack.Push(rootEnum)
            Do
                Dim childEnum = childStack.Peek

                If childEnum.MoveNext() Then
                    Dim child = childEnum.Current
                    childStack.Push(child.GetChildElements.GetEnumerator)
                Else
                    childStack.Pop()
                    Yield childStack.Peek.Current.Value
                End If
            Loop Until childStack.Count = 1
        End Function

        <Extension>
        Public Iterator Function BreadthTraverse(Of T)(tn As ITreeElement(Of T)) As IEnumerable(Of T)
            Dim queue As New Queue(Of ITreeElement(Of T))

            queue.Enqueue(tn)
            While Not queue.Count = 0
                tn = queue.Dequeue()
                Yield tn.Value
                For Each child In tn.GetChildElements
                    queue.Enqueue(child)
                Next
            End While
        End Function


        <Extension>
        Public Iterator Function DepthPreorderTraverseWithLevel(Of T)(tn As ITreeElement(Of T)) As IEnumerable(Of Tuple(Of T, Integer))
            Dim childStack As New Stack(Of IEnumerator(Of ITreeElement(Of T)))

            childStack.Push(tn.GetChildElements.GetEnumerator)
            Yield Tuple.Create(tn.Value, childStack.Count - 1)
            Do Until childStack.Count = 0
                Dim childEnum = childStack.Peek

                If childEnum.MoveNext() Then
                    Dim child = childEnum.Current
                    childStack.Push(child.GetChildElements.GetEnumerator)
                    Yield Tuple.Create(child.Value, childStack.Count - 1)
                Else
                    childStack.Pop()
                End If
            Loop

        End Function

        <Extension>
        Public Iterator Function DepthPostorderTraverseWithLevel(Of T)(tn As ITreeElement(Of T)) As IEnumerable(Of Tuple(Of T, Integer))
            Dim childStack As New Stack(Of IEnumerator(Of ITreeElement(Of T)))
            Dim rootEnum = {tn}.ToList.GetEnumerator
            childStack.Push(rootEnum)
            Do
                Dim childEnum = childStack.Peek

                If childEnum.MoveNext() Then
                    Dim child = childEnum.Current
                    childStack.Push(child.GetChildElements.GetEnumerator)
                Else
                    childStack.Pop()
                    Yield Tuple.Create(childStack.Peek.Current.Value, childStack.Count - 1)
                End If
            Loop Until childStack.Count = 1
        End Function

        <Extension>
        Private Iterator Function DepthPostorderTraverseWithLevel(Of T)(tn As ITreeElement(Of T), level As Integer) As IEnumerable(Of Tuple(Of T, Integer))
            For Each child In tn.GetChildElements
                For Each result In child.DepthPostorderTraverseWithLevel(level + 1)
                    Yield result
                Next
            Next
            Yield Tuple.Create(tn.Value, level)
        End Function


        <Extension>
        Public Iterator Function BreadthTraverseWithLevel(Of T)(tn As ITreeElement(Of T)) As IEnumerable(Of Tuple(Of T, Integer))
            Dim queue As New Queue(Of ITreeElement(Of T))
            Dim levelQueue As New Queue(Of Integer)

            queue.Enqueue(tn)
            levelQueue.Enqueue(0)
            While Not queue.Count = 0
                tn = queue.Dequeue()
                Dim level = levelQueue.Dequeue
                Yield Tuple.Create(tn.Value, level)
                For Each child In tn.GetChildElements
                    queue.Enqueue(child)
                    levelQueue.Enqueue(level + 1)
                Next
            End While
        End Function

    End Module
End Namespace

File: Src/Icm.Core/Tree/ITreeElementExtensions.cs
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;

namespace Icm.Tree
{

	/// <summary>
	/// Tree traversals.
	/// </summary>
	/// <remarks>
	/// 
	/// </remarks>
	public static class ITreeElementExtensions
	{
		public static IEnumerable<T> DepthPreorderTraverse<T>(this ITreeElement<T> tn)
		{
			Stack<IEnumerator<ITreeElement<T>>> childStack = new Stack<IEnumerator<ITreeElement<T>>>();

			childStack.Push(tn.GetChildElements().GetEnumerator());
			yield return tn.Value;
			while (childStack.Count != 0) {
				var childEnum = childStack.Peek();

				if (childEnum.MoveNext()) {
					var child = childEnum.Current;
					childStack.Push(child.GetChildElements().GetEnumerator());
					yield return child.Value;
				} else {
					childStack.Pop();
				}
			}
		}

		public static IEnumerable<T> DepthPostorderTraverse<T>(this ITreeElement<T> tn)
		{
			Stack<IEnumerator<ITreeElement<T>>> childStack = new Stack<IEnumerator<ITreeElement<T>>>();
			var rootEnum = new[] { tn }.ToList().GetEnumerator();
			childStack.Push(rootEnum);
			do {
				var childEnum = childStack.Peek();

				if (childEnum.MoveNext()) {
					var child = childEnum.Current;
					childStack.Push(child.GetChildElements().GetEnumerator());
				} else {
					childStack.Pop();
					yield return childStack.Peek().Current.Value;
				}
			} while (childStack.Count != 1);
		}
        
		public static IEnumerable<T> BreadthTraverse<T>(this ITreeElement<T> tn)
		{
			Queue<ITreeElement<T>> queue = new Queue<ITreeElement<T>>();

			queue.Enqueue(tn);
			while (queue.Count != 0) {
				tn = queue.Dequeue();
				yield return tn.Value;
				foreach (var child in tn.GetChildElements()) {
					queue.Enqueue(child);
				}
			}
		}

		public static IEnumerable<TraverseResult<T>> DepthPreorderTraverseWithLevel<T>(this ITreeElement<T> tn)
		{
			Stack<IEnumerator<ITreeElement<T>>> childStack = new Stack<IEnumerator<ITreeElement<T>>>();

			childStack.Push(tn.GetChildElements().GetEnumerator());
			yield return TraverseResult.Create(tn.Value, childStack.Count - 1);
			while (childStack.Count != 0) {
				var childEnum = childStack.Peek();

				if (childEnum.MoveNext()) {
					var child = childEnum.Current;
					childStack.Push(child.GetChildElements().GetEnumerator());
					yield return TraverseResult.Create(child.Value, childStack.Count - 1);
				} else {
					childStack.Pop();
				}
			}

		}
        
		public static IEnumerable<TraverseResult<T>> DepthPostorderTraverseWithLevel<T>(this ITreeElement<T> tn)
		{
			Stack<IEnumerator<ITreeElement<T>>> childStack = new Stack<IEnumerator<ITreeElement<T>>>();
			var rootEnum = new[] { tn }.ToList().GetEnumerator();
			childStack.Push(rootEnum);
			do {
				var childEnum = childStack.Peek();

				if (childEnum.MoveNext()) {
					var child = childEnum.Current;
					childStack.Push(child.GetChildElements().GetEnumerator());
				} else {
					childStack.Pop();
					yield return TraverseResult.Create(childStack.Peek().Current.Value, childStack.Count - 1);
				}
			} while (childStack.Count != 1);
		}
        
		private static IEnumerable<TraverseResult<T>> DepthPostorderTraverseWithLevel<T>(this ITreeElement<T> tn, int level)
		{
		    foreach (var result in tn.GetChildElements().SelectMany(child => child.DepthPostorderTraverseWithLevel(level + 1)))
		    {
		        yield return result;
		    }

		    yield return TraverseResult.Create(tn.Value, level);
		}

		public static IEnumerable<TraverseResult<T>> BreadthTraverseWithLevel<T>(this ITreeElement<T> tn)
		{
			Queue<ITreeElement<T>> queue = new Queue<ITreeElement<T>>();
			Queue<int> levelQueue = new Queue<int>();

			queue.Enqueue(tn);
			levelQueue.Enqueue(0);
			while (queue.Count != 0) {
				tn = queue.Dequeue();
				var level = levelQueue.Dequeue();
				yield return TraverseResult.Create(tn.Value, level);
				foreach (var child in tn.GetChildElements()) {
					queue.Enqueue(child);
					levelQueue.Enqueue(level + 1);
				}
			}
		}

	}
}